### PR TITLE
Partial retry

### PR DESCRIPTION
**Pull request by rpmoore**
Initial pull request to get feedback

### CODE REVIEW

**Comment by shabtaisharon** [comment]
The code looks good to me.
Do we also handle the case of more than one fail on the same object?

File: Ds3/Helpers/FileHelpers.cs [hunk 1]
```diff
@@ -138,7 +138,7 @@ private static string ConvertPathToKey(string path)
         ///
         /// For use with ListObjectsForDirectory
         /// </summary>
-        /// <param name="path">Full path< too file/param>
+        /// <param name="path">Full pathtoo file</param>
```
Comment (shabtaisharon) [inline Ds3/Helpers/FileHelpers.cs]: missing a space between "path too"

File: Ds3/Helpers/Transferrers/PartialDataTransferrerDecorator.cs [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+           string objectName,
+           long blobOffset,
+           Guid jobId,
+           IEnumerable<Range> ranges,
+           Stream stream) 
+        {
+
+            var currentTry = 0;
+            var transferrer = _transferrer;
+            var _ranges = ranges;
```
Comment (shabtaisharon) [inline Ds3/Helpers/Transferrers/PartialDataTransferrerDecorator.cs]: why do we need 2 new vars for _transferrer and ranges?

File: Ds3/Helpers/Transferrers/PartialDataTransferrerDecorator.cs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+                {
+                    transferrer.Transfer(client, bucketName, objectName, blobOffset, jobId, _ranges, stream);
+                    break;
+                }
+                catch (Ds3ContentLengthNotMatch exception)
+                {
+
+                    if (_retries != -1 && currentTry >= _retries)
+                    {
+                        throw new Ds3NoMoreRetriesException("Exhausted retries for retrieving data when partial data was received.", exception, currentTry);
```
Comment (shabtaisharon) [inline Ds3/Helpers/Transferrers/PartialDataTransferrerDecorator.cs]: put this string in the Resources

File: TestDs3/Helpers/Jobs/TestJobsUtil.cs [hunk 4]
```diff
@@ -0,0 +1,111 @@
+﻿using System;
```
Comment (shabtaisharon) [inline TestDs3/Helpers/Jobs/TestJobsUtil.cs]: missing Copyright header

File: Ds3/Helpers/Ds3ClientHelpers.cs [hunk 5]
```diff
@@ -32,11 +32,13 @@ public class Ds3ClientHelpers : IDs3ClientHelpers
         private const string JobTypePut = "PUT";
         private const string JobTypeGet = "GET";
         private readonly int _retryAfter; //-1 represent infinite number
+        private readonly int _getObjectRetries; // -1 represents infintie number of retries
 
-        public Ds3ClientHelpers(IDs3Client client, int retryAfter = -1)
+        public Ds3ClientHelpers(IDs3Client client, int retryAfter = -1, int getObjectRetries = 5)
```
Comment (berkeleysquare) [inline Ds3/Helpers/Ds3ClientHelpers.cs]: Publish the default as static DEFAULT_OBJECT_RETRIES = 5 ?

File: Ds3/Helpers/Jobs/PartialReadJob.cs [hunk 6]
```diff
@@ -30,7 +30,8 @@ internal class PartialReadJob : Job<IPartialReadJob, Ds3PartialObject>, IPartial
             JobResponse jobResponse,
             IEnumerable<string> fullObjects,
             IEnumerable<Ds3PartialObject> partialObjects,
-            ITransferItemSource transferItemSource)
+            ITransferItemSource transferItemSource,
+            int getObjectRetries = 5)
```
Comment (berkeleysquare) [inline Ds3/Helpers/Jobs/PartialReadJob.cs]: Like Ds3ClientHelpers: publish the default as static DEFAULT_OBJECT_RETRIES = 5 ?

**Pull request merged**